includes/amp-helper-functions.php

```diff
@@ -406,17 +406,33 @@ function amp_print_schemaorg_metadata() {
  *
  * @since 0.7.0
  */
-function amp_prepare_comment_post() {
-	if ( ! isset( $_GET['__amp_source_origin'] ) ) { // WPCS: CSRF ok. Beware of AMP_Theme_Support::purge_amp_query_vars().
+function amp_prepare_xhr_post() {
+	global $pagenow;
+	if ( ! isset( $_GET['__amp_source_origin'] ) || ! isset( $pagenow ) ) { // WPCS: CSRF ok. Beware of AMP_Theme_Support::purge_amp_query_vars().
 		return;
 	}
 
-	// Add amp comment hooks.
-	add_filter( 'comment_post_redirect', function() {
-		// We don't need any data, so just send a success.
-		wp_send_json_success();
-	}, PHP_INT_MAX, 2 );
-
+	if ( 'wp-comments-post.php' === $pagenow ) {
+		// This only runs on wp-comments-post.php.
+		add_filter( 'comment_post_redirect', function() {
+			// We don't need any data, so just send a success.
+			wp_send_json_success();
+		}, PHP_INT_MAX, 2 );
+	} elseif ( ! isset( $_GET['_wp_amp_action_xhr_converted'] ) ) { // WPCS: CSRF ok.
+		// submission was from a set action-xhr, implying it's being handled already.
+		return;
+	} else {
+		// Add amp redirect hooks.
+		add_filter( 'wp_redirect', 'amp_handle_general_post', PHP_INT_MAX, 2 );
+		add_action( 'template_redirect', function() {
+			/*
+			 * Buffering starts here, so unlikely the form has a redirect,
+			 * so force a redirect to the same page.
```

Comment:
This is clever, but I'm worried it could have unintended results. In particular, when a form submission does not do a redirect after a `POST` then the data from the `POST` submission is often displayed in some way on the response. If we do the redirect, then it could cause the user to be redirected back to a blank page. Consider something like this:

```php
<form method="post">
     <?php if ( ! empty( $_POST ) ) : ?>
          Submission successful!
          <pre><?php echo esc_html( wp_json_encode( wp_unslash( $_POST ) ) ); ?></pre>
     <?php else : ?>
         <input name="data"><button>Submit</button>
     <?php endif; ?>
</form>
```

Something like this will break if we do the redirect here.